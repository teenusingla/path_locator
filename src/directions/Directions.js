import React, { Component } from "react";
import { DirectionsForm, DirectionsMap, RouteInfo } from "./components";
import { fetchDirections } from "./services";
import { Loader } from "../common/loader";

import "./Directions.scss";

/**
 * @type {Component}
 * @name Directions
 * @description Directions components (Main view of directions module)
 */
class Directions extends Component {
  /**
   * @constructor
   */
  constructor() {
    super();

    // initial state
    this.state = {
      directionsResponse: null,
      isLoading: false
    };
  }

  /**
   * @description Fetch the directions
   * This method call the Directions serive to fetch the data
   * @param from Starting Location
   * @param to Drop-off point
   */
  getDirections = async (from, to) => {
    this.toggleLoader(true);
    // directions result, if error then show error messgae
    const response = await fetchDirections(from, to).catch(e => {
      this.showErrorMessage("Internal server error");
    });

    this.toggleLoader(false);

    // check for response error
    if (response && response.error) {
      this.showErrorMessage(response.error);
      return;
    }

    // check for directions data, if available then set the data to state
    if (response && response.path) {
      this.setState(() => ({
        directionsResponse: response
      }));
    }
  };

  /**
   * @description Hide and show loader
   * @param isLoading Boolean indicating that loader is shown or not
   */
  toggleLoader = isLoading => {
    this.setState(() => ({
      isLoading
    }));
  };

  /**
   * @name showErrorMessage
   * @description This method display the error message
   * @param message Message to be displayed
   */
  showErrorMessage = message => {
    this.toggleLoader(false);
    this.setState(() => ({
      directionsResponse: null
    }));
    alert(message);
  };

  /**
   * @description Render method of the component
   */
  render() {
    const { directionsResponse, isLoading } = this.state;
    return (
      <div className="directions-container">
        <Loader isLoading={isLoading} />
        <div className="direction-form-container">
          <DirectionsForm getDirections={this.getDirections} />
          <div className="directions-route-info">
            {directionsResponse && <RouteInfo {...directionsResponse} />}
          </div>
        </div>
        <div className="direction-map-container">
          <DirectionsMap directions={directionsResponse} />
        </div>
      </div>
    );
  }
}

export default Directions;
