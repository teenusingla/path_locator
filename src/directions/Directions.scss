.directions-container {
  width: 100%;
  height: 100%;
}

.directions-container::after {
  content: "";
  display: block;
  clear: both;
}

.direction-map-container {
  height: 100%;
  width: 75%;
  position: absolute;
  left: 25%;
}

.direction-form-container {
  position: absolute;
  left: 0;
  width: 25%;
  padding: 20px;
  height: 100%;
}
