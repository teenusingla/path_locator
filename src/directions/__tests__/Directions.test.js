import React from "react";
import renderer from "react-test-renderer";
import Directions from "../Directions";

describe("<Directions /> component", () => {
  it("should match the snapshot", () => {
    const tree = renderer.create(<Directions />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
