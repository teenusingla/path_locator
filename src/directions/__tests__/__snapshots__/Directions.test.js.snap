// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Directions /> component should match the snapshot 1`] = `
<div
  className="directions-container"
>
  <div
    className="direction-form-container"
  >
    <div
      className="directions-form"
    >
      <div
        className="rows"
      >
        <p>
          Start Location
        </p>
        <input
          className="input-map"
          name="startLocation"
          onChange={[Function]}
          type="text"
        />
      </div>
      <div
        className="rows"
      >
        <p>
          Drop off Location
        </p>
        <input
          className="input-map"
          name="dropLocation"
          onChange={[Function]}
          type="text"
        />
      </div>
      <div
        className="form-control"
      >
        <button
          className="btn"
          onClick={[Function]}
        >
          Submit
        </button>
        <button
          className="btn"
          onClick={[Function]}
        >
          Reset
        </button>
      </div>
    </div>
    <div
      className="directions-route-info"
    />
  </div>
  <div
    className="direction-map-container"
  >
    <div
      className="map-container"
    >
      <div />
    </div>
  </div>
</div>
`;
